include/Player/WaterGun.hpp

```diff
@@ -0,0 +1,285 @@
+            "/mario/watergun2/back_wg",                // mPath
+            "/mario/watergun2/back_wg/back_wg.bmd",    // mBmdPath
+            1,                                          // mNumEmitters
+            2,                                          // _15
+            {
+                {1, 0},
+                {4, 0},
+                {4, 0}
+            }
+        }
+    }
+};
+
+class TWaterGun {
+public:
+    typedef u8 TNozzleType;
+    enum {
+        Spray = 0,
+        Rocket,
+        Underwater,
+        Yoshi,
+        Hover,
+        Turbo
+    };
+
+    TWaterGun(TMario *mario);
+
+    virtual void perform(u32, JDrama::TGraphics *);
+
+    void calcAnimation(JDrama::TGraphics *);
+    void changeBackup();
+    void changeNozzle(TNozzleType, bool);
+    bool damage();
+    void emit();
+    TNozzleBase *getCurrentNozzle() const {
+        return this->mNozzleList[this->mCurrentNozzle];
+    }
+    MtxPtr getEmitMtx(int);
+    void getEmitPosDirSpeed(int, JGeometry::TVec3<f32> *pos, JGeometry::TVec3<f32> *dir, JGeometry::TVec3<f32> *speed);
+    MtxPtr getNozzleMtx();
+    f32 getPressure();
+    f32 getPressureMax();
+    void init();
+    void initInLoadAfter();
+    bool isEmitting();
+    bool isPressureOn();
+    void movement();
+    void rotateProp(f32);
+    void setAmountToRate(f32);
+    void setBaseTRMtx(Mtx);
+    bool suck();
+    void startDashEffect();
+    void endDashEffect();
+    void getMarioUpperStatus();
+
+    inline MtxPtr getYoshiMtx() {
+        TYoshi *yoshi = (TYoshi*)this->mMario->yoshi;
+        return yoshi->mActor->mModel->getAnmMtx(yoshi->mJoint);
+    }
+    
+    inline MtxPtr getWillBeEmitted(int jointIndex) {
+
+        // This entire block is likely an inlined function.
+        u8 currentNozzle = this->mCurrentNozzle;
+        s8 flag = nozzleBmdData.getFlags(currentNozzle, jointIndex);
+        if (flag < 3) {
+            return this->getCurrentNozzle()->mActor->mModel->getAnmMtx(nozzleBmdData.getJointIndex(this->mCurrentNozzle, jointIndex));
+        } else if (flag == 3) {
+            return getYoshiMtx();
+        }
+    }
+
+    void decWater() {
+    }
+
+    bool hasWater() const {
+        return this->mCurrentWater > 0;
+    }
+
+    u32 getWaterGunAnmId();
+    
+    J3DModel *getModel() {
+        return this->mFluddModel->mModel;
+    }
+
+    inline void setEmitPt() {
+        int index;
+        int offset;
+        // TODO: Register r27 and r28 are swapped on some instructions. I cannot figure out how to swap them correctly.
+        for (index = 0, offset = 0; index < nozzleBmdData.getEmitterCount(this->mCurrentNozzle); index += 1, offset += 0xc) {
+            MtxPtr p1 = this->getEmitMtx(index);
+            if (p1 != nullptr) {
+                *(float *)((int)&this->mEmitPos[0].x + offset) = (p1)[0][3];
+                *(float *)((int)&this->mEmitPos[0].y + offset) = (p1)[1][3];
+                *(float *)((int)&this->mEmitPos[0].z + offset) = (p1)[2][3];
```

Comment:
This doesn't look like something they would've written. I think it should be
```
mEmitPos[index].x = p1[0][3];
mEmitPos[index].y = p1[1][3];
mEmitPos[index].z = p1[2][3];
```
which might also fix the regswap

Reply:
I don't disagree that this is 99% wrong, but it is the closest i could get. Doing it like suggested was further off than a register diff. Not sure exactly why, but i'm suspecting that this is some kind of inline setter as well. I will add a comment about it!

Reply:
I think it's better to leave it matching less but making more sense. I've seen analogous problems in a couple of other places, so once we have enough examples, it should be easier to figure out the real inline (which is probably inside of TVec<f32>). But I'm not insisting, it's okay to leave as-is too =)